### PR TITLE
[archive] Add introduction to map-reduce

### PR DESCRIPTION
**Pull request by nullist0**
### Motivation

The map-reduce model is very popular from 2004 when the paper is posted. I studied it and this will help us to understand the Map-reduce model.

### Modifications

- Add Map-reduce markdown files.

### Result

- Added introduction of Map-reduce implementations.

### CODE REVIEW

File: archive/map-reduce.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+## Model definition
+
+The programming model can be defined as the input and output of its execution. There are three execution model on Map-reduce model, that are the full program, map function and reduce function.
+
+- Full execution
+  - Input: a set of key/value pair
+  - Output: a set of key/value pair
+- Map: the function of producing intermediate key/value pair
+  - Input: a key/value pair
```
Comment (TheWeaVer) [inline archive/map-reduce.md:26]:
```suggestion
  - Input: a set of key/value pair
```

File: archive/map-reduce.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  - Output: a set of key/value pair
```
Comment (TheWeaVer) [inline archive/map-reduce.md:27]: maybe...? 
```suggestion
  - Output: a set of intermediate key/value pairs
```

File: archive/map-reduce.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+- Reduce: the function of producing intermediate key/value pair
+  - Input: an intermediate key I and a set of values for that key.
+  - Output: the desired ouput on the given intermediate key I
+With pseudocode, we can write as below. The specific implementation needs to be written by the developers.
+
+```
+def program(input: Pair<K, V>): Output<K, O>
+  do 
+
+def map(input: Pair<K, V>): Set<Intermediates<T, W>>
```
Comment (TheWeaVer) [inline archive/map-reduce.md:37]:
```suggestion
def map(input: Pair<K, V>): Set<Intermediate<T, W>>
```

Comment (nullist0) [reply archive/map-reduce.md]: I applied this at 217e370f276880aa93ef23c1d833c48e533d6191.

File: archive/map-reduce.md (hunk 1 above)
Comment (nullist0) [reply archive/map-reduce.md:26]: From the paper, the map will accept an input pair.

image

Comment (TheWeaVer) [reply archive/map-reduce.md:26]: Oh, I see. Thank you. Then, why ... the program - Full execution - requires set?
```
- Full execution
  - Input: a set of key/value pair
  - Output: a set of key/value pair
```
https://github.com/TheWeaVer/lia/pull/26/files#diff-d6e6fe5544ed282cd2bd3e96db96991677b08a950ba2ca1ecc6f93553d8b1b9dR22-R24

Comment (nullist0) [reply archive/map-reduce.md:26]: The full execution is a computation. The right above lines of the above image. Let me change the `full execution` to `computation`. Sorry to make a confusion. 😢

Comment (TheWeaVer) [reply archive/map-reduce.md:26]: Ah, then I understand that the full procedure ! Thank you !

**Review by TheWeaVer — APPROVED** [review]
LGTM

File: archive/map-reduce.md (hunk 1 above)
Comment (nullist0) [reply archive/map-reduce.md:26]: This is applied at a3852f5afa4573109179c9debebc43562dcf1c97.

**Pull request merged**